File: survival-client/src/main/java/net/survival/client/ColumnMaskFactory.java
package net.survival.client;

import java.util.HashSet;
import java.util.Set;

import net.survival.block.ColumnPos;

public final class ColumnMaskFactory {
    private ColumnMaskFactory() {}

    public static Set<Long> createCircle(int radius, double offsetX, double offsetZ) {
        var radiusSquared = radius * radius;
        var offsetCX = ColumnPos.toColumnX((int) Math.floor(offsetX));
        var offsetCZ = ColumnPos.toColumnZ((int) Math.floor(offsetZ));
        var mask = new HashSet<Long>((int) Math.ceil(Math.PI * radiusSquared));

        for (var z = -radius; z < radius; ++z) {
            for (var x = -radius; x < radius; ++x) {
                if (squareDistance(x, z) <= radiusSquared) {
                    var cx = offsetCX + x;
                    var cz = offsetCZ + z;
                    mask.add(ColumnPos.hashPos(cx, cz));
                }
            }
        }

        return mask;
    }

    private static int squareDistance(int x, int z) {
        return (x * x) + (z * z);
    }
}

File: survival-client/src/main/java/net/survival/client/ColumnMaskFactory.kt
package net.survival.client

import net.survival.block.ColumnPos
import java.util.*
import kotlin.math.ceil
import kotlin.math.floor

object ColumnMaskFactory {
    fun createCircle(radius: Int, offsetX: Double, offsetZ: Double): Set<Long> {
        val radiusSquared = radius * radius
        val offsetCX = ColumnPos.toColumnX(floor(offsetX).toInt())
        val offsetCZ = ColumnPos.toColumnZ(floor(offsetZ).toInt())
        val mask = HashSet<Long>(ceil(Math.PI * radiusSquared).toInt())
        for (z in -radius until radius) {
            for (x in -radius until radius) {
                if (squareDistance(x, z) <= radiusSquared) {
                    val cx = offsetCX + x
                    val cz = offsetCZ + z
                    mask.add(ColumnPos.hashPos(cx, cz))
                }
            }
        }
        return mask
    }

    private fun squareDistance(x: Int, z: Int): Int {
        return x * x + z * z
    }
}